File: 19_控制预测类_一元线性回归分析/Regress.py
import numpy as np  # 导入 numpy 库用于数值计算
import statsmodels.api as sm  # 导入 statsmodels 库用于回归分析
import matplotlib.pyplot as plt  # 导入 matplotlib 库用于绘图

# 1. 输入数据
# 输入 X 的样本值
x = np.array([143, 145, 146, 147, 149, 150, 153, 154, 155, 156, 157, 158, 159, 160, 162, 164])
# 插入 β0 对应列（截距项）
X = sm.add_constant(x)  # 在 x 中添加常数列（截距项），用于表示 β0
# 输入 Y 的样本值
Y = np.array([88, 85, 88, 91, 92, 93, 93, 95, 96, 98, 97, 96, 98, 99, 100, 102])

# 2. 回归分析及检验
model = sm.OLS(Y, X)  # 定义线性回归模型，使用最小二乘法（OLS）
results = model.fit()  # 拟合模型，计算回归系数等统计量
# model.fit()函数返回一个包含回归分析结果的对象，包括拟合参数、标准误差、t检验值、p值等。

# 输出我们需要的数据
print("回归系数 (b):")  # 输出回归系数
print(results.params)  # results.params 包含截距项 β0 和斜率 β1(Matlab中的b--回归系数)
print("回归系数的置信区间 (bint):")  # 输出回归系数的置信区间
print(results.conf_int())  # results.conf_int() 返回每个系数的置信区间(Matlab中的bint)
print("统计量 (stats):")
print("R-squared:", results.rsquared)  # 输出 R-squared 值，表示模型的解释力(Matlab中的r,拟合优度)
print("F-statistic:", results.fvalue)  # 输出 F 统计量，用于检验模型的整体显著性(Matlab中的F)
print("p-value:", results.f_pvalue)  # 输出 p 值，用于判断模型是否显著(Matlab中的p)
print("Residual variance:", results.mse_resid)  # 输出残差方差
print("Residual standard error:", np.sqrt(results.mse_resid))  # 输出残差标准误差
print(results.summary())    # 输出回归分析结果的摘要信息

# 3. 残差分析，做残差图
residuals = results.resid  # 获取残差（实际值与预测值的差异）
residual_std_error = np.sqrt(results.mse_resid)  # 计算残差的标准误差

fig, ax = plt.subplots()  # 创建绘图对象       subplots()函数返回一个包含 figure 和 axes 对象的元组
ax.axhline(y=0, color='k', linestyle='--')  # 添加零线，表示残差为零的位置
# 绘制残差及其置信区间
ax.errorbar(x, residuals, yerr=residual_std_error, fmt='o', capsize=5)  
#fmt表示点的形状，capsize表示误差线的宽度
ax.set_xlabel('X')  # 设置 x 轴标签
ax.set_ylabel('Residuals')  # 设置 y 轴标签
ax.set_title('Residual Plot')  # 设置图表标题
plt.show()  # 显示残差图

# 4. 预测及作图
z = results.predict(X)  # 使用拟合的模型进行预测，得到预测值
plt.plot(x, Y, 'k+', label='实际值')  # 绘制实际值，使用黑色加号标记
plt.plot(x, z, 'r-', label='拟合值')  # 绘制拟合值，使用红色实线
plt.xlabel('X')  # 设置 x 轴标签
plt.ylabel('Y')  # 设置 y 轴标签
plt.title('回归分析结果')  # 设置图表标题
plt.legend()  # 添加图例
plt.show()  # 显示回归分析结果图


# result.summary()  # 输出回归分析结果的摘要信息
# result.params  # 回归系数                         (matlab中的b,表示斜率)
# result.conf_int()  # 回归系数的置信区间            (matlab中的bint,表示斜率的置信区间)
# result.rsquared  # R-squared 值                  (拟合优度,表示模型的解释力)
# result.fvalue  # F 统计量                         (F值越大，线性关系越显著)
# result.f_pvalue  # p 值                             (p值越小，线性关系越显著)
# result.mse_resid  # 残差方差                       (残差方差越小，残差越接近正态分布)
# np.sqrt(result.mse_resid)  # 残差标准误差          (残差标准误差越小，残差越接近正态分布)
# result.resid  # 残差                             (实际值与预测值的差异)
# result.predict(X)  # 预测值                         (使用拟合的模型进行预测)

File: 19_控制预测类_一元线性回归分析/Regress.m
%% 1、输入数据
%输入X的样本值
clear;clc;
x=[143 145 146 147 149 150 153 154 155 156 157 158 159 160 162 164]'; 
%插入\beta0对应列
X=[ones(16,1) x];
%输入Y的样本值
Y=[88 85 88 91 92 93 93 95 96 98 97 96 98 99 100 102]';

%% 2、回归分析及检验：
[b,bint,r,rint,stats]=regress(Y,X);    % regress函数用于回归分析

%regress(Y,X)函数返回值说明：
% b:回归系数，即回归直线的斜率
% bint:回归系数的置信区间
% r:R-squared，即回归模型的拟合优度
% rint:R-squared的置信区间
% stats:回归分析的统计量(决定系数)，包括F值、p值、无偏估计等
% 决定系数 :越接近1，拟合优度越好；越接近0，拟合优度越差。
% F值 : F值越大，线性关系越显著
% p值 : p值越小，拟合优度越好；p值越大，拟合优度越差。(F值对应的犯错的概率P值)
% 残差的方差

%输出我们需要的数据
b,bint,stats
%结果：
% b =
%   -16.0730
%     0.7194
% bint =
%   -33.7071    1.5612
%     0.6047    0.8340
% stats =
%     0.9282(决定系数)  180.9531(F值)    0.0000(P值)    1.7437(残差的标准误差)
% 即β_0=-16.073 ,β_1=0.7194; 
% β_0 的置信区间为[-33.7017 ,1.5612]，β_1 的置信区间为[0.6047 ,0.834]; 
% r_2=0.9282,F=180.9531,p=0.0000 
% 由p<0.05,可知回归模型 y=-16.073+0.7194x 成立

%% 3、残差分析，做残差图
rcoplot(r,rint)   %rcplot(r,rint)函数用于残差图
%rcoplot(r,rint)函数返回值说明：
% r:残差值
% rint:残差值的置信区间
% 图中红色的点为异常值，蓝色的点为正常值
% 图中红色的线为残差的置信区间，蓝色的线为残差的平均值
% 图中红色的点越多，说明异常值越多，模型的拟合优度越差
% 图中蓝色的点越多，说明正常值越多，模型的拟合优度越好
% 图中红色的线越长，说明残差的置信区间越大，模型的拟合优度越差
% 图中蓝色的线越短，说明残差的平均值越小，模型的拟合优度越好

%从残差图可以看出，除第二个数据外，其余数据的残差离
% 零点均较近，且残差的置信区间均包含零点，这说明回归模型
% y=-16.073+0.7194x能较好的符合原始数据，而第二个数据可视为异常点

% 4、预测及作图
z=b(1)+b(2)*x 
plot(x,Y,'k+',x,z,'r')

